File: Leetcode/PHP/3. Longest Substring Without Repeating Characters.php
<?php
class Solution{
    function lengthOfLongestSubstring($s) {
        $n = strlen($s);
        $ans = 0;
        $map = [];
        for($i = 0, $j = 0; $j < $n; $j++){
            if(array_key_exists($s[$j], $map)){
                $i = max($map[$s[$j]], $i);
            }
            $ans = max($ans, $j - $i + 1);
            $map[$s[$j]] = $j + 1;
        }
        return $ans;
    }
}

$s = new Solution();
echo $s->lengthOfLongestSubstring("abcabcbb");
echo $s->lengthOfLongestSubstring("bbbbb");
echo $s->lengthOfLongestSubstring("pwwkew");

?>

File: Leetcode/C++/3_Longest_Substring_Without_Repeating_Characters.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
    public:
        int lengthOfLongestSubstring(string s) {
            int n = s.size();
            if (n == 0) return 0;
            int ans = 1;
            int start = 0;
            unordered_map<char, int> last;
            last[s[0]] = 0;
            for (int i = 1; i < n; i++) {
                if (last.find(s[i]) != last.end() && last[s[i]] >= start) {
                    start = last[s[i]] + 1;
                }
                last[s[i]] = i;
                ans = max(ans, i - start + 1);
            }
            return ans;
        }
};

int main() {
    Solution s;
    cout << s.lengthOfLongestSubstring("abcabcbb") << endl;
    cout << s.lengthOfLongestSubstring("bbbbb") << endl;
    cout << s.lengthOfLongestSubstring("pwwkew") << endl;
    return 0;
}